### ISSUE TITLE
Does not execute on Pico Max using latest SDK and Ubuntu image on EMMC/SD

### ISSUE THREAD

**Issue by platima** [luckfox-eng29/luckfox_pico_rtsp_yolov5#2]
```
pico@luckfox:~$ ls
luckfox_rtsp_yolov5_ubuntu
pico@luckfox:~$ chmod a+x luckfox_rtsp_yolov5_ubuntu
pico@luckfox:~$ ./luckfox_rtsp_yolov5_ubuntu
-bash: ./luckfox_rtsp_yolov5_ubuntu: No such file or directory
pico@luckfox:~$ ldd luckfox_rtsp_yolov5_ubuntu
        librknnmrt.so => not found
        librockiva.so => not found
        librockit.so => not found
        librockchip_mpp.so.1 => not found
        librkaiq.so => not found
        librga.so => not found
        libstdc++.so.6 => /lib/arm-linux-gnueabihf/libstdc++.so.6 (0xa6d80000)
        libgcc_s.so.1 => /lib/arm-linux-gnueabihf/libgcc_s.so.1 (0xa6d50000)
        libc.so.0 => not found
        libm.so.6 => /lib/arm-linux-gnueabihf/libm.so.6 (0xa6cf0000)
        libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xa6bc0000)
        /lib/ld-uClibc.so.0 => /lib/ld-linux-armhf.so.3 (0xa6f0b000)
```

**Comment by luckfox-eng29** [comment]
Ubuntu uses glibc as its standard C library, but the RKMPI libraries provided by Rockchip are built based on uClibc. This causes linking issues in the Ubuntu environment, preventing them from running correctly. Rockchip has not provided the source code for RKMPI and RKNN, so I am unable to build a version suitable for Ubuntu. I apologize for not mentioning these details in the documentation. I have now updated it accordingly.

**Comment by platima** [comment]
Could you not statically link uClibc to the binary?

I tried copying it into the same directory, and confirmed `ldd` picked it up locally, but it would not launch properly.

I also tried modifying the SDK to statically compile everything so there were no dynamic libraries required, but this also did not work.

A lot of customers / YouTube viewers are wanting to run the demos under the Ubuntu environment, so there must be some solution like this?

**Comment by luckfox-eng29** [comment]
I initially considered using static linking for this project, but unfortunately, after testing, I found that it wasn't successful. The ability to call RKNN and RKMPI is limited to the closed-source libraries provided by Rockchip. Without access to the source code, everything becomes quite challenging.

Additionally, there are issues with using Ubuntu, such as the camera not being able to capture images properly. It can only read RAW data but cannot pass it to RKISP for processing. To run on the Ubuntu platform, we would need to address the RKISP library issue.

We also aim to leverage Ubuntu's robust software ecosystem to enhance the usability of Luckfox-pico. However, it seems that the feasibility of adapting Ubuntu using static linking is low at the moment. Personally, I lean towards using containers to tackle this problem, segregating the Ubuntu environments with glibc and uclibc. However, my lack of experience with containers leaves me uncertain about the viability of this approach.

**Comment by platima** [comment]
Yep I had the same issue when I tried static linking. Makes sense re the libraries, but that is also a pain!

I did not realise that with the Ubuntu images. Also makes sense.

Containers would be an interesting approach, but would quite possibly cause some other incompatibilities, plus there are performance overheads.

It's a shame Alpine moved away from uClibc, but I do understand their reasoning. RKNN does't support musl? No distribution really supports uClibc any more from what I can see.

THAT BEING SAID, I just found that in https://github.com/airockchip/rknn-toolkit2/tree/master/rknpu2/runtime/Linux/rknn_server there are now two versions for armhf, one that looks to use glibc;
```
owner@NB08:~/rknn-toolkit2/rknpu2/runtime/Linux/rknn_server$ file armhf*/usr/bin/rknn_server
armhf-uclibc/usr/bin/rknn_server: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-uClibc.so.0, stripped
armhf/usr/bin/rknn_server:        ELF 32-bit LSB executable, ARM, EABI5 version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, stripped
```

So perhaps using this binary would work on the Ubuntu image? I am not on a computer with the SDK / qemu / Docker to test right now, but that update from 2 months ago for ARM 32 bit looks perfect for Ubuntu on Cortex-A7 like the RV110X.

**Comment by luckfox-eng29** [comment]
According to Rockchip's documentation, the armhf-related libraries and executable files in rknn_server are intended for use with the RK3xxx series of chips. While Rockchip's NPUs are all designed based on the TIM-VX scheme, different NPUs may have different hardware scheduling mechanisms. I have not attempted to call armhf-glibc libraries, but the anticipated results are not optimistic

**Comment by platima** [comment]
Hrmm, you may be right but I am still tryign to be optimistic. I am guessing you were referring to https://github.com/rockchip-linux/rknpu2/blob/master/rknn_server_proxy.md? (I just translated to EN and read it)

I can see librknnrt.so is meant for RK3 and librknnmrt.so for RK1, and that does match up with the files;
```
./rknpu2/runtime/Linux/librknn_api/armhf/librknnrt.so
./rknpu2/runtime/Linux/librknn_api/armhf-uclibc/librknnmrt.so
```

I think it might be worth trying with the glibc SO, both static and dynamic, just in case. I would not expect that the RK API would expose different symbols, maybe the RV110X with reduced symbols/functions, but that should mean that the RK3XXX version works, is just bigger or slower? (Coming from a software developer background).

It's a shame they are stripped.

**Comment by luckfox-eng29** [comment]
hat's an interesting thought. While Rockchip may not have designed for compatibility, for the sake of cost-effectiveness in the RV1106 series, they might not have made significant changes in NPU and software design. Using the glibc library may result in only partial functionality (specific to the RK3xxx series) being unavailable. I'll give it a try.

**Comment by platima** [comment]
Possibly. I might try too if I get time this weekend. I really want RKNN working under Ubuntu, or even Alpine linux!

Good luck.
